File: TarsyShared/Sources/TarsyShared/Networking/WSProtocol.swift
import Foundation

public enum WSAction: String, Codable, Sendable {
    // Workspace
    case workspaceList = "workspace:list"
    case workspaceCreate = "workspace:create"
    case workspaceStart = "workspace:start"
    case workspaceStop = "workspace:stop"
    case workspaceStatus = "workspace:status"
    case workspaceUpdate = "workspace:update"
    case workspaceScanRepos = "workspace:scan_repos"
    case workspaceScanResult = "workspace:scan_result"

    // Stream
    case streamStart = "stream:start"
    case streamStop = "stream:stop"
    case streamFrame = "stream:frame"

    // Remote Input
    case remoteTap = "remote:tap"
    case remoteDoubleTap = "remote:double_tap"
    case remoteLongPress = "remote:long_press"
    case remoteScroll = "remote:scroll"
    case remoteDrag = "remote:drag"
    case remoteScrollStart = "remote:scroll_start"
    case remoteScrollEnd = "remote:scroll_end"
    case remotePinch = "remote:pinch"
    case remotePinchStart = "remote:pinch_start"
    case remotePinchEnd = "remote:pinch_end"
    case remoteKeyboard = "remote:keyboard"
    case remoteButton = "remote:button"
    case screenshotRequest = "screenshot:request"
    case screenshotResult = "screenshot:result"

    // Dev Server
    case devServerStart = "devserver:start"
    case devServerStop = "devserver:stop"
    case devServerStatus = "devserver:status"

    // Browser
    case browserOpenUrl = "browser:open_url"
    case browserBack = "browser:back"
    case browserForward = "browser:forward"
    case browserRefresh = "browser:refresh"
    case browserMobileViewport = "browser:mobile_viewport"
    case browserDesktopViewport = "browser:desktop_viewport"
    case browserTabList = "browser:tab_list"
    case browserTabListResult = "browser:tab_list_result"
    case browserTabSwitch = "browser:tab_switch"
    case browserTabClose = "browser:tab_close"

    // HTTP Proxy (WKWebView tunnel)
    case proxyDetectPorts = "proxy:detect_ports"
    case proxyDetectPortsResult = "proxy:detect_ports_result"
    case proxyRequest = "proxy:request"
    case proxyResponse = "proxy:response"

    // Terminal
    case terminalCreate = "terminal:create"
    case terminalInput = "terminal:input"
    case terminalOutput = "terminal:output"
    case terminalClose = "terminal:close"
    case terminalList = "terminal:list"
    case terminalComplete = "terminal:complete"
    case terminalCompleteResult = "terminal:complete_result"

    // Claude Code
    case claudeCreate = "claude:create"
    case claudeMessage = "claude:message"
    case claudeOutput = "claude:output"
    case claudeComplete = "claude:complete"
    case claudeClose = "claude:close"
    case claudeAskUser = "claude:ask_user"
    case claudeUserResponse = "claude:user_response"

    // OpenClaw
    case openclawStatus = "openclaw:status"
    case openclawMessage = "openclaw:message"
    case openclawOutput = "openclaw:output"
    case openclawComplete = "openclaw:complete"

    // Generic Engine (multi-provider)
    case engineCreate = "engine:create"
    case engineMessage = "engine:message"
    case engineOutput = "engine:output"
    case engineComplete = "engine:complete"
    case engineClose = "engine:close"
    case engineError = "engine:error"
    case engineAskUser = "engine:ask_user"
    case engineUserResponse = "engine:user_response"

    // Git
    case gitCheckpoint = "git:checkpoint"
    case gitCheckpointResult = "git:checkpoint_result"
    case gitDiff = "git:diff"
    case gitDiffResult = "git:diff_result"
    case gitRollback = "git:rollback"
    case gitRollbackResult = "git:rollback_result"
    case gitHistory = "git:history"
    case gitHistoryResult = "git:history_result"
    case gitFileDiff = "git:file_diff"
    case gitFileDiffResult = "git:file_diff_result"
    case gitBranches = "git:branches"
    case gitBranchesResult = "git:branches_result"
    case gitCheckout = "git:checkout"
    case gitCheckoutResult = "git:checkout_result"
    case gitPull = "git:pull"
    case gitPullResult = "git:pull_result"
    case gitStage = "git:stage"
    case gitStageResult = "git:stage_result"
    case gitDiscard = "git:discard"
    case gitDiscardResult = "git:discard_result"

    // File Explorer
    case fileTree = "file:tree"
    case fileTreeResult = "file:tree_result"
    case fileRead = "file:read"
    case fileReadResult = "file:read_result"

    // Engine Status (model, tokens, context %)
    case engineStatus = "engine:status"

    // MCP Store
    case mcpList = "mcp:list"
    case mcpListResult = "mcp:list_result"
    case mcpHealthCheck = "mcp:health_check"
    case mcpHealthResult = "mcp:health_result"

    // Relay
    case relayMachineOnline = "relay:machine_online"
    case relayMachineOffline = "relay:machine_offline"
    case relayStreamFrame = "relay:stream_frame"
    case relayNoClients = "relay:no_clients"

    // Sudo
    case sudoRequest = "sudo:request"
    case sudoResponse = "sudo:response"
    case sudoResult = "sudo:result"

    // Agent Detection
    case agentsDetected = "agents:detected"
    case slashCommandsRequest = "agents:slash_commands_request"
    case slashCommandsDetected = "agents:slash_commands"

    // Agent Settings
    case agentSettings = "agent:settings"
    case agentSettingsUpdate = "agent:settings_update"

    // Repo Analysis
    case repoAnalyze = "repo:analyze"
    case repoAnalysis = "repo:analysis"

    // AI Project Wizard
    case wizardStart = "wizard:start"
    case wizardResponse = "wizard:response"
    case wizardExecute = "wizard:execute"
    case wizardResult = "wizard:result"
    case wizardGhDetected = "wizard:gh_detected"

    // UltraContext
    case ultracontextStatus = "ultracontext:status"

    // DevTools
    case processList = "devtools:process_list"
    case processListResult = "devtools:process_list_result"
    case processKill = "devtools:process_kill"
    case processKillResult = "devtools:process_kill_result"
    case portsList = "devtools:ports_list"
    case portsListResult = "devtools:ports_list_result"
    case httpRequest = "devtools:http_request"
    case httpResponse = "devtools:http_response"
    case systemResources = "devtools:system_resources"
    case systemResourcesResult = "devtools:system_resources_result"

    // Security
    case securityRotateSecret = "security:rotate_machine_secret"
    case securityRotateResult = "security:rotate_result"
    case securityFingerprintUpdate = "security:fingerprint_update"
    case e2eEncrypted = "e2e:encrypted"
    case e2eKeyExchange = "e2e:key_exchange"
    case e2eKeyExchangeResponse = "e2e:key_exchange_response"

    // System
    case auth
    case authSuccess = "auth:success"
    case authFail = "auth:fail"
    case ping
    case pong
    case error
}

public struct WSPacket: Codable, Sendable {
    public let id: String
    public let action: WSAction
    public let payload: [String: String]?
    public let timestamp: Date

    public init(action: WSAction, payload: [String: String]? = nil, id: String = UUID().uuidString) {
        self.id = id
        self.action = action
        self.payload = payload
        self.timestamp = Date()
    }

    public func encode() throws -> Data {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return try encoder.encode(self)
    }

    /// Maximum allowed size for a single WSPacket payload in bytes (1MB).
    /// Binary frames (video, screenshots) bypass this limit as they use raw WebSocket binary messages.
    private static let maxPacketSize = 1_048_576

    public static func decode(from data: Data) throws -> WSPacket {
        guard data.count <= maxPacketSize else {
            throw WSPacketError.payloadTooLarge(data.count)
        }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(WSPacket.self, from: data)
    }

    public enum WSPacketError: LocalizedError {
        case payloadTooLarge(Int)

        public var errorDescription: String? {
            switch self {
            case .payloadTooLarge(let size):
                return "WSPacket too large: \(size) bytes (max \(WSPacket.maxPacketSize))"
            }
        }
    }
}

File: TarsyWindows/Networking/WSProtocol.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace TarsyWindows.Networking;

/// <summary>
/// All WebSocket protocol actions — mirrors WSProtocol.swift.
/// </summary>
public static class WSAction
{
    // System
    public const string Auth = "auth";
    public const string AuthSuccess = "auth:success";
    public const string AuthFail = "auth:fail";
    public const string Ping = "ping";
    public const string Pong = "pong";
    public const string Error = "error";

    // Workspace
    public const string WorkspaceList = "workspace:list";
    public const string WorkspaceCreate = "workspace:create";
    public const string WorkspaceStart = "workspace:start";
    public const string WorkspaceStop = "workspace:stop";
    public const string WorkspaceStatus = "workspace:status";
    public const string WorkspaceUpdate = "workspace:update";
    public const string WorkspaceScanRepos = "workspace:scan_repos";
    public const string WorkspaceScanResult = "workspace:scan_result";

    // Stream
    public const string StreamStart = "stream:start";
    public const string StreamStop = "stream:stop";
    public const string StreamFrame = "stream:frame";

    // Remote Input
    public const string RemoteTap = "remote:tap";
    public const string RemoteDoubleTap = "remote:double_tap";
    public const string RemoteLongPress = "remote:long_press";
    public const string RemoteScroll = "remote:scroll";
    public const string RemoteDrag = "remote:drag";
    public const string RemoteScrollStart = "remote:scroll_start";
    public const string RemoteScrollEnd = "remote:scroll_end";
    public const string RemotePinch = "remote:pinch";
    public const string RemotePinchStart = "remote:pinch_start";
    public const string RemotePinchEnd = "remote:pinch_end";
    public const string RemoteKeyboard = "remote:keyboard";
    public const string RemoteButton = "remote:button";
    public const string ScreenshotRequest = "screenshot:request";
    public const string ScreenshotResult = "screenshot:result";

    // Dev Server
    public const string DevServerStart = "devserver:start";
    public const string DevServerStop = "devserver:stop";
    public const string DevServerStatus = "devserver:status";

    // Browser
    public const string BrowserOpenUrl = "browser:open_url";
    public const string BrowserBack = "browser:back";
    public const string BrowserForward = "browser:forward";
    public const string BrowserRefresh = "browser:refresh";
    public const string BrowserMobileViewport = "browser:mobile_viewport";
    public const string BrowserDesktopViewport = "browser:desktop_viewport";
    public const string BrowserTabList = "browser:tab_list";
    public const string BrowserTabListResult = "browser:tab_list_result";
    public const string BrowserTabSwitch = "browser:tab_switch";
    public const string BrowserTabClose = "browser:tab_close";

    // HTTP Proxy
    public const string ProxyDetectPorts = "proxy:detect_ports";
    public const string ProxyDetectPortsResult = "proxy:detect_ports_result";
    public const string ProxyRequest = "proxy:request";
    public const string ProxyResponse = "proxy:response";

    // Terminal
    public const string TerminalCreate = "terminal:create";
    public const string TerminalInput = "terminal:input";
    public const string TerminalOutput = "terminal:output";
    public const string TerminalClose = "terminal:close";
    public const string TerminalList = "terminal:list";
    public const string TerminalComplete = "terminal:complete";
    public const string TerminalCompleteResult = "terminal:complete_result";
    public const string TerminalInterrupt = "terminal:interrupt";

    // Engine Interrupt
    public const string EngineInterrupt = "engine:interrupt";

    // Claude Code
    public const string ClaudeCreate = "claude:create";
    public const string ClaudeMessage = "claude:message";
    public const string ClaudeOutput = "claude:output";
    public const string ClaudeComplete = "claude:complete";
    public const string ClaudeClose = "claude:close";
    public const string ClaudeAskUser = "claude:ask_user";
    public const string ClaudeUserResponse = "claude:user_response";

    // OpenClaw
    public const string OpenclawStatus = "openclaw:status";
    public const string OpenclawMessage = "openclaw:message";
    public const string OpenclawOutput = "openclaw:output";
    public const string OpenclawComplete = "openclaw:complete";

    // Generic Engine (multi-provider)
    public const string EngineCreate = "engine:create";
    public const string EngineMessage = "engine:message";
    public const string EngineOutput = "engine:output";
    public const string EngineComplete = "engine:complete";
    public const string EngineClose = "engine:close";
    public const string EngineError = "engine:error";
    public const string EngineAskUser = "engine:ask_user";
    public const string EngineUserResponse = "engine:user_response";

    // Git
    public const string GitCheckpoint = "git:checkpoint";
    public const string GitCheckpointResult = "git:checkpoint_result";
    public const string GitDiff = "git:diff";
    public const string GitDiffResult = "git:diff_result";
    public const string GitRollback = "git:rollback";
    public const string GitRollbackResult = "git:rollback_result";
    public const string GitHistory = "git:history";
    public const string GitHistoryResult = "git:history_result";
    public const string GitFileDiff = "git:file_diff";
    public const string GitFileDiffResult = "git:file_diff_result";
    public const string GitBranches = "git:branches";
    public const string GitBranchesResult = "git:branches_result";
    public const string GitCheckout = "git:checkout";
    public const string GitCheckoutResult = "git:checkout_result";
    public const string GitPull = "git:pull";
    public const string GitPullResult = "git:pull_result";
    public const string GitStage = "git:stage";
    public const string GitStageResult = "git:stage_result";
    public const string GitDiscard = "git:discard";
    public const string GitDiscardResult = "git:discard_result";

    // File Explorer
    public const string FileTree = "file:tree";
    public const string FileTreeResult = "file:tree_result";
    public const string FileRead = "file:read";
    public const string FileReadResult = "file:read_result";

    // Engine Status
    public const string EngineStatus = "engine:status";

    // MCP Store
    public const string McpList = "mcp:list";
    public const string McpListResult = "mcp:list_result";
    public const string McpHealthCheck = "mcp:health_check";
    public const string McpHealthResult = "mcp:health_result";

    // Relay
    public const string RelayMachineOnline = "relay:machine_online";
    public const string RelayMachineOffline = "relay:machine_offline";
    public const string RelayStreamFrame = "relay:stream_frame";
    public const string RelayNoClients = "relay:no_clients";

    // Sudo / UAC
    public const string SudoRequest = "sudo:request";
    public const string SudoResponse = "sudo:response";
    public const string SudoResult = "sudo:result";

    // Agent Detection
    public const string AgentsDetected = "agents:detected";
    public const string SlashCommandsRequest = "agents:slash_commands_request";
    public const string SlashCommandsDetected = "agents:slash_commands";

    // Agent Settings
    public const string AgentSettings = "agent:settings";
    public const string AgentSettingsUpdate = "agent:settings_update";

    // Repo Analysis
    public const string RepoAnalyze = "repo:analyze";
    public const string RepoAnalysis = "repo:analysis";

    // AI Project Wizard
    public const string WizardStart = "wizard:start";
    public const string WizardResponse = "wizard:response";
    public const string WizardExecute = "wizard:execute";
    public const string WizardResult = "wizard:result";
    public const string WizardGhDetected = "wizard:gh_detected";

    // UltraContext
    public const string UltracontextStatus = "ultracontext:status";

    // DevTools
    public const string ProcessList = "devtools:process_list";
    public const string ProcessListResult = "devtools:process_list_result";
    public const string ProcessKill = "devtools:process_kill";
    public const string ProcessKillResult = "devtools:process_kill_result";
    public const string PortsList = "devtools:ports_list";
    public const string PortsListResult = "devtools:ports_list_result";
    public const string HttpRequest = "devtools:http_request";
    public const string HttpResponse = "devtools:http_response";
    public const string SystemResources = "devtools:system_resources";
    public const string SystemResourcesResult = "devtools:system_resources_result";

    // Security
    public const string SecurityRotateSecret = "security:rotate_machine_secret";
    public const string SecurityRotateResult = "security:rotate_result";
    public const string SecurityFingerprintUpdate = "security:fingerprint_update";
    public const string E2eEncrypted = "e2e:encrypted";
    public const string E2eKeyExchange = "e2e:key_exchange";
    public const string E2eKeyExchangeResponse = "e2e:key_exchange_response";
}

public record WSPacket
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = Guid.NewGuid().ToString();

    [JsonPropertyName("action")]
    public string Action { get; init; } = "";

    [JsonPropertyName("payload")]
    public Dictionary<string, string>? Payload { get; init; }

    [JsonPropertyName("timestamp")]
    public DateTime Timestamp { get; init; } = DateTime.UtcNow;

    public static WSPacket Create(string action, Dictionary<string, string>? payload = null)
    {
        return new WSPacket
        {
            Id = Guid.NewGuid().ToString(),
            Action = action,
            Payload = payload,
            Timestamp = DateTime.UtcNow,
        };
    }

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public string Encode()
    {
        var json = JsonSerializer.Serialize(this, JsonOptions);
        if (json.Length > MaxPacketSize)
            throw new InvalidOperationException($"WSPacket too large: {json.Length} bytes (max {MaxPacketSize})");
        return json;
    }

    public static WSPacket Decode(string json)
    {
        if (json.Length > MaxPacketSize)
            throw new InvalidOperationException($"WSPacket too large: {json.Length} bytes (max {MaxPacketSize})");
        return JsonSerializer.Deserialize<WSPacket>(json, JsonOptions)
            ?? throw new InvalidOperationException("Failed to decode WSPacket");
    }

    private const int MaxPacketSize = 1_048_576; // 1 MB
}
